File: Sources/SwiftFrontend/Ball.swift
import Foundation
import simd

struct BallWorld {
  var balls: [Ball] = []

  mutating func add(position: SIMD2<Float>, angle: Float, size: Float) {
    balls.append(Ball(position: position, angle: angle, size: size))
  }

  mutating func update(_ deltaTime: Float) {
    for i in balls.indices {
      balls[i].update(deltaTime)
    }
  }
}

struct Ball {
  public static let speed: Float = 80.0
  public static let worldSize: Float = 512.0

  var _position: SIMD2<Float>, _velocity: SIMD2<Float>
  var _size: Float

  init(position: SIMD2<Float>, angle: Float, size ballSize: Float) {
    _position = position
    _velocity = SIMD2<Float>(
       cos(angle * Float.pi * 2.0),
      -sin(angle * Float.pi * 2.0)
    )
    _size = ballSize
  }

  public mutating func update(_ deltaTime: Float) {
    _position += _velocity * Self.speed * deltaTime
    if (_position.x < _size) {
      _velocity.x = -_velocity.x
      _position.x = _size
    } else if (_position.x > Self.worldSize - _size) {
      _velocity.x = -_velocity.x
      _position.x = Self.worldSize - _size
    }
    if (_position.y < _size) {
      _velocity.y = -_velocity.y
      _position.y = _size
    } else if (_position.y > Self.worldSize - _size) {
      _velocity.y = -_velocity.y
      _position.y = Self.worldSize - _size
    }
  }

  var position: SIMD2<Float> { return _position }
  var velocity: SIMD2<Float> { return _velocity }
  var size: Float { return _size }
}

File: Sources/CppBackend/ball.cpp
#include "ball.hpp"
#include <cmath>


Ball::Ball(vec2f pos, float angle, float ballSize) noexcept :
  _position(pos),
  _velocity(simd::make<vec2f>(
     std::cos(angle * M_PI * 2.0f),
    -std::sin(angle * M_PI * 2.0f)
  )),
  _size(ballSize) {
}

void Ball::update(float deltaTime) noexcept {
  _position += _velocity * speed * deltaTime;
  if (_position.x < _size) {
    _velocity.x = -_velocity.x;
    _position.x = _size;
  } else if (_position.x > worldSize - _size) {
    _velocity.x = -_velocity.x;
    _position.x = worldSize - _size;
  }
  if (_position.y < _size) {
    _velocity.y = -_velocity.y;
    _position.y = _size;
  } else if (_position.y > worldSize - _size) {
    _velocity.y = -_velocity.y;
    _position.y = worldSize - _size;
  }
}


void BallWorld::add(Ball::vec2f pos, float angle, float ballSize) noexcept {
  balls.emplace_back(Ball{ pos, angle, ballSize });
}

void BallWorld::update(float deltaTime) noexcept {
  for (auto& ball : balls) {
    ball.update(deltaTime);
  }
}
